lib/puppet_x/counsyl/pip.rb

```diff
@@ -44,7 +44,7 @@ def install
             # is specified to ensure package is actually installed.
             self.class.instances.each do |pip_package|
               if pip_package.name == pypkg
-                args << '--upgrade'
+                args << '--upgrade --upgrade-strategy only-if-needed'
```

Comment:
Would a check like this work? @lucaswiman 
```ruby
if args.include? "upgrade-strategy"
  args << '--upgrade'
else
  args << '--upgrade --upgrade-strategy only-if-needed'
end

Reply:
I think `args` should be a list of strings, each of which is an argument. I'd expect `pip` to raise an error that `--upgrade --upgrade-strategy only-if-needed only-if-needed` isn't a valid option. I think you'd want is the following, though I'd suggest testing it out:
```ruby
args << '--upgrade'
if !args.include? "upgrade-strategy"
  args += ['--upgrade-strategy', 'only-if-needed']
end
```

Reply:
Wait, no, `!args.include? "--upgrade-strategy"`:
```
irb(main):012:0> ['--upgrade-strategy'].include? 'upgrade-strategy'
=> false
```

Reply:
Thanks @lucaswiman, isn't the format of install_options a list of dictionaries though?
```
[{'--upgrade-strategy' => 'only-if-needed'}]
```
Does that get turned into a list of strings via some puppetry magic, I was looking around and didn't see anything obvious.

Reply:
I thought based on the code you wrote that it was a list of strings. If it's a list of hashes, then I don't see how your code does what you want:
```
irb(main):004:0> [{'--upgrade-strategy' => 'only-if-needed'}].include? 'upgrade-strategy'
=> false
```

Reply:
Apparently it can be a list of strings or hashes (args and kwargs) according to the [puppet-spec](https://puppet.com/docs/puppet/5.3/type.html#package-attribute-install_options)

Reply:
Alright, how about this:

```ruby
if args.include? "--upgrade-strategy" or args.any? { |arg| arg.is_a?(Hash) and arg.key?("--upgrade-strategy")}
```

Reply:
lgtm